src/rai/rai/tools/ros/manipulation.py

```diff
@@ -0,0 +1,191 @@
+                "/manipulator_move_to",
+            ),
+        )
+
+    def _run(
+        self,
+        x: float,
+        y: float,
+        z: float,
+        task: Literal["grab", "place"],
+    ) -> str:
+        pose_stamped = PoseStamped()
+        pose_stamped.header.frame_id = self.manipulator_frame
+        pose_stamped.pose = Pose(
+            position=Point(x=x, y=y, z=z),
+            orientation=self.quaternion,
+        )
+
+        if task == "place":
+            pose_stamped.pose.position.z += self.additional_height
+
+        pose_stamped.pose.position.x += self.calibration_x
+        pose_stamped.pose.position.y += self.calibration_y
+        pose_stamped.pose.position.z += self.calibration_z
+
+        pose_stamped.pose.position.z = np.max(
+            [pose_stamped.pose.position.z, self.min_z]
+        )  # avoid hitting the table
+
+        request = ManipulatorMoveTo.Request()
+        request.target_pose = pose_stamped
+
+        if task == "grab":
+            request.initial_gripper_state = True  # open
+            request.final_gripper_state = False  # closed
+        else:
+            request.initial_gripper_state = False  # closed
+            request.final_gripper_state = True  # open
+
+        future = self.client.call_async(request)
+        self.node.get_logger().debug(
+            f"Calling ManipulatorMoveTo service with request: x={request.target_pose.pose.position.x:.2f}, y={request.target_pose.pose.position.y:.2f}, z={request.target_pose.pose.position.z:.2f}"
+        )
+
+        rclpy.spin_until_future_complete(self.node, future)
+
+        if future.result() is not None:
+            response = future.result()
+            if response.success:
+                return f"End effector successfully positioned at coordinates ({x:.2f}, {y:.2f}, {z:.2f}). Note: The status of object interaction (grab/drop) is not confirmed by this movement."
+            else:
+                return f"Failed to position end effector at coordinates ({x:.2f}, {y:.2f}, {z:.2f})."
+        else:
+            return f"Service call failed for point ({x:.2f}, {y:.2f}, {z:.2f})."
+
+
+class GetObjectPositionsToolInput(BaseModel):
+    object_name: str = Field(
+        ..., description="The name of the object to get the positions of"
+    )
+
+
+class GetObjectPositionsTool(BaseTool):
+    name: str = "get_object_positions"
+    description: str = (
+        "Retrieve the positions of all objects of a specified type within the manipulator's frame of reference. "
+        "This tool provides accurate positional data but does not distinguish between different colors of the same object type. "
+        "While position detection is reliable, please note that object classification may occasionally be inaccurate."
+    )
+
+    target_frame: str  # frame of the manipulator
+    source_frame: str  # frame of the camera
+    camera_topic: str  # rgb camera topic
+    depth_topic: str
+    camera_info_topic: str  # rgb camera info topic
```

Comment:
There is a number of assumptions in this code that makes manipulation feature more of a PoC, please make sure this is documented as such